### ISSUE TITLE
Incorrect decoding of SavedJobMessages

### ISSUE THREAD

**Issue by JoshuaCarter** [symbiote/silverstripe-queuedjobs#228]
Issue is present on the tip of 3.x and 4.x versions.

### Outline:

- Run an iterative job (lets call it 'MyJob').
- 'MyJob' job *doesn't* populate `SavedJobData` with any data, thus `N;` (serialized null) is written into the DB.
- 'MyJob' also populates `SavedJobMessages` with data.
- 'MyJob' completes it's first `process` tick.
- QueuedJobService begins preparing for the second tick of 'MyJob'.
- `copyDescriptorToJob` deserializes data and messages (see https://github.com/symbiote/silverstripe-queuedjobs/blob/7a989c37016fb4cd79a14ccc8226e5c1db051765/src/Services/QueuedJobService.php#L268).
- `SavedJobMessages` is deserialized correctly into an array of messages.
- Because `SavedJobData` is just `N;`, the output is just `null`.
- Because `$jobData` is null, **BOTH** data and messages are overridden with the output from `json_decode`.
- Running `json_decode` on serialized php (primarily the messages) returns `null`.
- Thus message data is lost.

I may throw a PR at this in the near future. I would want to target 3.x and 4.x. I should target '3' and 'master' branches, right?

**Comment by robbieaverill** [comment]
>I should target '3' and 'master' branches, right?

If you can fix this in a non-breaking way, then please target one of the patch release branches, e.g. 3.0 or 3.1 and 4.0, 4.1, 4.2 or 4.3. Note that we can easily merge up the commit if you target the lowest release version e.g. 3.0 and 4.0 =)

**Comment by JoshuaCarter** [comment]
@robbieaverill I've thrown up 2 PRs for this here:
- [3.0 PR](https://github.com/symbiote/silverstripe-queuedjobs/pull/229)
- [4.0 PR](https://github.com/symbiote/silverstripe-queuedjobs/pull/230)

**Comment by robbieaverill** [comment]
>The only thing worth noting is that the removal of Convert::json2obj() should be entirely without consequence as all it's ever done is a json_decode (as far as I could find at least).

Yep, that's fine. That method is also deprecated in favour of using json_decode directly.

**Comment by JoshuaCarter** [comment]
@robbieaverill PRs updated, thanks for reviewing, they should be good to go (but don't take my word for it) :)

**Comment by JoshuaCarter** [comment]
@robbieaverill Champion.
